File: apps/web/lib/api/ai-entry.ts
import { publicEnv } from "@/lib/config";
import type { AiSuggestion, ConfirmAiPostingResponse } from "@/lib/api/ai-accountant";

export type AiEntryWorkbenchResponse = {
  suggestion: AiSuggestion;
  workflow_state: string;
  confidence_band: "high" | "medium" | "low";
  clarification_questions: string[];
  extracted_invoice: {
    vendor_or_customer_name: string | null;
    invoice_number: string | null;
    invoice_date: string | null;
    gstin: string | null;
    taxable_amount: string | null;
    cgst_amount: string | null;
    sgst_amount: string | null;
    igst_amount: string | null;
    total_amount: string | null;
    source_text_available: boolean;
    extraction_warning: string | null;
  } | null;
  doctor_findings: string[];
  gst_risks: string[];
  duplicate_warnings: string[];
};

export type AiAccuracyDashboard = {
  total_suggestions: number;
  approved_suggestions: number;
  approved_without_edit: number;
  rejected_suggestions: number;
  corrected_suggestions: number;
  approval_rate: string;
  correction_rate: string;
  estimated_accuracy: string;
  average_confidence: string;
};

export type AiMonthEndReadiness = {
  books_completion_percent: string;
  pending_vouchers: number;
  unreconciled_bank_entries: number;
  gst_risk_count: number;
  missing_bill_count: number;
  readiness_score: string;
};

export type AiOwnerReport = {
  month: string;
  summary: string;
  profit: string;
  cash_position: string;
  gst_payable: string;
};

export type AiCorrection = {
  id: string;
  suggestion_id: string | null;
  corrected_payload: Record<string, unknown>;
  created_at: string;
};

async function api<T>(path: string, token: string, options: { method?: string; body?: unknown } = {}): Promise<T> {
  let response: Response;
  try {
    response = await fetch(`${publicEnv.NEXT_PUBLIC_API_URL}${path}`, {
      method: options.method ?? "GET",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json"
      },
      body: options.body === undefined ? undefined : JSON.stringify(options.body)
    });
  } catch {
    throw new Error(`API not reachable. Check NEXT_PUBLIC_API_URL. Current API: ${publicEnv.NEXT_PUBLIC_API_URL}`);
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(typeof error.detail === "string" ? error.detail : "Request failed");
  }

  return response.json() as Promise<T>;
}

export const aiEntryApi = {
  inbox: (companyId: string, token: string, text: string, sourceType: string) =>
    api<AiEntryWorkbenchResponse>(`/companies/${companyId}/ai-entry/inbox`, token, {
      method: "POST",
      body: { text, source_type: sourceType }
    }),
  parseText: (companyId: string, token: string, text: string) =>
    api<AiEntryWorkbenchResponse>(`/companies/${companyId}/ai-entry/parse-text`, token, {
      method: "POST",
      body: { text, source_type: "one_line_text" }
    }),
  uploadPdf: (companyId: string, token: string, filename: string, fileBase64: string) =>
    api<AiEntryWorkbenchResponse>(`/companies/${companyId}/ai-entry/upload-pdf`, token, {
      method: "POST",
      body: { filename, file_base64: fileBase64 }
    }),
  approve: (companyId: string, token: string, suggestionId: string) =>
    api<ConfirmAiPostingResponse>(`/companies/${companyId}/ai-entry/approve`, token, {
      method: "POST",
      body: { suggestion_id: suggestionId }
    }),
  reject: (companyId: string, token: string, suggestionId: string, reason: string) =>
    api<AiEntryWorkbenchResponse>(`/companies/${companyId}/ai-entry/reject`, token, {
      method: "POST",
      body: { suggestion_id: suggestionId, reason }
    }),
  accuracy: (companyId: string, token: string) =>
    api<AiAccuracyDashboard>(`/companies/${companyId}/ai-entry/accuracy-dashboard`, token),
  corrections: (companyId: string, token: string) =>
    api<AiCorrection[]>(`/companies/${companyId}/ai-entry/corrections`, token),
  readiness: (companyId: string, token: string) =>
    api<AiMonthEndReadiness>(`/companies/${companyId}/ai-entry/month-end-readiness`, token),
  ownerReport: (companyId: string, token: string) =>
    api<AiOwnerReport>(`/companies/${companyId}/ai-entry/owner-report`, token)
};

File: apps/api/app/schemas/ai_entry.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.accounting import AiSuggestionStatus, VoucherType
from app.schemas.ai_accountant import AiVoucherSuggestionResponse
from app.schemas.accounting import VoucherResponse


class AiEntryTextRequest(BaseModel):
    text: str = Field(min_length=3, max_length=1200)
    transaction_date: date | None = None
    language: str = "auto"
    source_type: str = "one_line_text"


class AiEntryPdfRequest(BaseModel):
    filename: str = Field(min_length=1, max_length=180)
    file_base64: str = Field(min_length=10, max_length=14_000_000)
    transaction_date: date | None = None
    language: str = "auto"


class AiExtractedInvoiceFields(BaseModel):
    vendor_or_customer_name: str | None = None
    invoice_number: str | None = None
    invoice_date: date | None = None
    gstin: str | None = None
    taxable_amount: Decimal | None = None
    cgst_amount: Decimal | None = None
    sgst_amount: Decimal | None = None
    igst_amount: Decimal | None = None
    total_amount: Decimal | None = None
    source_text_available: bool
    extraction_warning: str | None = None


class AiEntrySuggestionResponse(BaseModel):
    suggestion: AiVoucherSuggestionResponse
    workflow_state: str
    confidence_band: str
    clarification_questions: list[str]
    extracted_invoice: AiExtractedInvoiceFields | None = None
    doctor_findings: list[str] = []
    gst_risks: list[str] = []
    duplicate_warnings: list[str] = []


class AiEntrySuggestionListItem(BaseModel):
    suggestion_id: UUID
    input_text: str
    status: AiSuggestionStatus
    confidence: Decimal
    voucher_type: VoucherType
    amount: Decimal
    created_at: str
    decided_at: str | None


class AiEntryApproveRequest(BaseModel):
    suggestion_id: UUID
    corrected_payload: dict | None = None
    correction_note: str | None = Field(default=None, max_length=500)


class AiEntryRejectRequest(BaseModel):
    suggestion_id: UUID
    reason: str = Field(min_length=2, max_length=500)


class AiEntryApproveResponse(BaseModel):
    suggestion_id: UUID
    voucher: VoucherResponse
    correction_recorded: bool


class AiCorrectionResponse(BaseModel):
    id: UUID
    suggestion_id: UUID | None
    corrected_payload: dict
    created_at: str


class AiAccuracyDashboardResponse(BaseModel):
    total_suggestions: int
    approved_suggestions: int
    approved_without_edit: int
    rejected_suggestions: int
    corrected_suggestions: int
    approval_rate: Decimal
    correction_rate: Decimal
    estimated_accuracy: Decimal
    average_confidence: Decimal


class AiMonthEndReadinessResponse(BaseModel):
    books_completion_percent: Decimal
    pending_vouchers: int
    unreconciled_bank_entries: int
    gst_risk_count: int
    missing_bill_count: int
    readiness_score: Decimal


class AiOwnerReportResponse(BaseModel):
    month: str
    summary: str
    profit: Decimal
    cash_position: Decimal
    gst_payable: Decimal
